### PR TITLE
Hp refinement

### PR DESCRIPTION
**Pull request by JPPayonk**
* add hp-refinement
* refactor adaptive refinement
* update mesh examples

### CODE REVIEW

File: ossdbs/api.py [hunk 1]
```diff
@@ -94,6 +95,19 @@ def generate_electrodes(settings: dict):
...
 
+        # Apply hp-refinement only on active contacts
+        if "HPRefinement" in settings["Mesh"]:
+            hp_settings = settings["Mesh"]["HPRefinement"]
+            if hp_settings["Active"] and "Contacts" in electrode_parameters:
+                for contact_info in electrode_parameters["Contacts"]:
+                    if contact_info["Active"]:
+                        contact_idx = contact_info["Contact_ID"]
+                        order = hp_settings["Order"]
+                        _set_edge_hp_flag(electrode, {f"Contact_{contact_idx}": order})
```
Comment (j-zimmermann) [inline ossdbs/api.py:106]: I would move this to the electrode class, this doesn't make sense in the API.
I.e, create a method and call it using `electrode.set_hp_flag` (should also be a public method)

File: ossdbs/api.py [hunk 2]
```diff
@@ -209,11 +223,9 @@ def generate_mesh(settings):
...
         return mesh
 
-    if "MeshingHypothesis" in mesh_settings:
-        mesh_hypothesis = mesh_settings["MeshingHypothesis"]
-    else:
-        mesh_hypothesis = {"Type": "Default"}
-    mesh.generate_mesh(mesh_hypothesis)
+    if "MeshingHypothesis" not in mesh_settings:
+        mesh_settings["MeshingHypothesis"] = {"Type": "Default"}
+    mesh.generate_mesh(mesh_settings)
```
Comment (j-zimmermann) [inline ossdbs/api.py:228]: What is the reasoning behind this change?

File: ossdbs/api.py [hunk 3]
```diff
@@ -619,3 +631,17 @@ def run_PAM(settings):
             scaling=settings["Scaling"],
             scaling_index=settings["ScalingIndex"],
         )
+
+
+def _set_edge_hp_flag(electrode: ElectrodeModel, edge_sizes: dict) -> None:
```
Comment (j-zimmermann) [inline ossdbs/api.py:636]: Move to electrode class as mentioned above

File: ossdbs/api.py [hunk 4]
```diff
@@ -619,3 +631,17 @@ def run_PAM(settings):
...
+    """Set flag on edges."""
+    for edge in electrode.geometry.edges:
+        if edge.name in edge_sizes:
+            edge.hpref = edge_sizes[edge.name]
+
+
+def _set_vertex_hp_flag(electrode: ElectrodeModel, vertex_sizes) -> None:
```
Comment (j-zimmermann) [inline ossdbs/api.py:643]: Move also this to electrode class

File: ossdbs/fem/mesh.py [hunk 5]
```diff
@@ -31,8 +34,14 @@ def __init__(self, geometry: netgen.occ.OCCGeometry, order: int) -> None:
 
     def generate_mesh(self, meshing_parameters: dict) -> None:
         """Generate NGSolve mesh."""
-        netgen_mp = self._meshing_parameters(meshing_parameters)
+        netgen_mp = self._meshing_parameters(meshing_parameters["MeshingHypothesis"])
```
Comment (j-zimmermann) [inline ossdbs/fem/mesh.py:37]: I'm not a fan of this change, what's the reason behind it?

File: ossdbs/fem/mesh.py [hunk 6]
```diff
@@ -31,8 +34,14 @@ def __init__(self, geometry: netgen.occ.OCCGeometry, order: int) -> None:
...
         self._mesh = ngsolve.Mesh(self.geometry.GenerateMesh(mp=netgen_mp))
+        if (
+            "HPRefinement" in meshing_parameters
```
Comment (j-zimmermann) [inline ossdbs/fem/mesh.py:40]: Is this condition safe? What happens if `HPRefinement` isn't in the dictionary?

File: ossdbs/fem/volume_conductor/volume_conductor_model.py [hunk 7]
```diff
@@ -244,7 +244,10 @@ def run_full_analysis(
                     self._impedances[band_indices] = impedance
 
                 # refine only at first frequency
-                if freq_idx == frequency_indices[0] and adaptive_mesh_refinement:
+                if (
+                    freq_idx == frequency_indices[0]
+                    and adaptive_mesh_refinement["Active"]
```
Comment (j-zimmermann) [inline ossdbs/fem/volume_conductor/volume_conductor_model.py:249]: Is this safe? Isn't `adaptive_mesh_refinement`  bool in most of the cases?

File: ossdbs/fem/volume_conductor/volume_conductor_model.py [hunk 8]
```diff
@@ -257,7 +260,11 @@ def run_full_analysis(
                     # currently: refine until impedance doesn't change by more than 0.1%
                     error = 100
                     refinements = 0
-                    while error > 0.1 and refinements < 10:
+                    # TODO write a meaningful algo
+                    while (
+                        error > 0.1
```
Comment (j-zimmermann) [inline ossdbs/fem/volume_conductor/volume_conductor_model.py:265]: Make the error also a parameter in the dictionary

**Review by j-zimmermann — CHANGES_REQUESTED** [review]
In general a nice PR, I'd suggest to fix the issues I raised and then it should be good to merge.

File: ossdbs/api.py (hunk 1 above)
Comment (JPPayonk) [reply ossdbs/api.py]: You are right; it is moved to the electrode_model_template.py as a public method and is called by ''electrode.set_hp_flag''

File: ossdbs/api.py (hunk 3 above)
Comment (JPPayonk) [reply ossdbs/api.py]: Done

File: ossdbs/api.py (hunk 4 above)
Comment (JPPayonk) [reply ossdbs/api.py]: Done

File: ossdbs/fem/mesh.py (hunk 6 above)
Comment (JPPayonk) [reply ossdbs/fem/mesh.py:40]: Yes, if `HPRefinement` is not in the dictionary, simply no hp-refinement is applied

File: ossdbs/fem/volume_conductor/volume_conductor_model.py (hunk 7 above)
Comment (JPPayonk) [reply ossdbs/fem/volume_conductor/volume_conductor_model.py:249]: Yes, the structure of the settings for adaptive mesh refinement has been slightly modified. Now it's a dictionary, where you can set it as ´True´ and ´False but also modify MaxIterations and the error.

File: ossdbs/fem/volume_conductor/volume_conductor_model.py (hunk 8 above)
Comment (JPPayonk) [reply ossdbs/fem/volume_conductor/volume_conductor_model.py]: Done

File: ossdbs/api.py (hunk 2 above)
Comment (JPPayonk) [reply ossdbs/api.py:224]: As the hp-refinement is called in the `mesh` class, we must provide the mesh_settings, not only the mesh_hypothesis.

File: ossdbs/fem/mesh.py (hunk 5 above)
Comment (JPPayonk) [reply ossdbs/fem/mesh.py:37]: As we now provide the mesh_settings to the method, we need to specify the Meshinghypothesis as the parameter for the `_meshing_parameters`.

**Review by j-zimmermann — CHANGES_REQUESTED** [review]

**Review by j-zimmermann — APPROVED** [review]

**Pull request merged**